tests/filecheck/dialects/pdl/verifiers.mlir

```diff
@@ -0,0 +1,86 @@
+// RUN: xdsl-opt %s --split-input-file --verify-diagnostics | filecheck %s
+
+pdl.pattern @NonConstantAttrInRewrite : benefit(0) {
+    %op = pdl.operation "test.op"
+    pdl.rewrite %op {
+        %attr = pdl.attribute
+        %new_op = pdl.operation "test.op" { "value" = %attr }
+        pdl.replace %op with %new_op
+    }
+}
+
+// CHECK: expected constant value when specified within a `pdl.rewrite`
+
+// -----
+
+pdl.pattern @UnnamedOpInRewrite : benefit(0) {
+    %op = pdl.operation "test.op"
+    pdl.rewrite %op {
+        %new_op = pdl.operation
+        pdl.replace %op with %new_op
+    }
+}
+
+// CHECK: must have an operation name when nested within a `pdl.rewrite`
+
+// -----
+
+pdl.pattern @UnusedAttribute : benefit(0) {
+    %attr = pdl.attribute = 0 : i32
+    %op = pdl.operation "test.op"
+    pdl.rewrite %op {
+        pdl.erase %op
+    }
+}
+
+// CHECK: expected a bindable user when defined in the matcher body of a `pdl.pattern`
+
+// -----
+
+pdl.pattern @UnusedType : benefit(0) {
+    %type = pdl.type : i32
+    %op = pdl.operation "test.op"
+    pdl.rewrite %op {
+        pdl.erase %op
+    }
+}
+
+// CHECK: expected a bindable user when defined in the matcher body of a `pdl.pattern`
+
+// -----
+
+pdl.pattern @UnusedTypes : benefit(0) {
+    %type = pdl.types : [i32, i32]
+    %op = pdl.operation "test.op"
+    pdl.rewrite %op {
+        pdl.erase %op
+    }
+}
+
+// CHECK: expected a bindable user when defined in the matcher body of a `pdl.pattern`
+
+// -----
+
+pdl.pattern @UnusedOperand : benefit(0) {
+    %type = pdl.type : i32
+    %operand = pdl.operand : %type
+    %op = pdl.operation "test.op"
+    pdl.rewrite %op {
+        pdl.erase %op
+    }
+}
+
+// CHECK: expected a bindable user when defined in the matcher body of a `pdl.pattern`
```

Comment:
My understanding is that tests with the same string are inherently not testing what you think they are, as the filecheck checks within the whole file, not the section within `// -----` delimiters. So actually three of the tests could "fail" but the filecheck would still pass. It would be good to add some identifiable information to these four tests that would ensure that they actually would signal failure if behaviour changed.

Reply:
Hmm, aren't they supposed to be printed only once?
Two `CHECK:` should not match the same line?

Reply:
My understanding is that the next `CHECK` can not be in the same line. I think in the current state a test could fail and it would not be apparent which one. 
If someone adds more tests to this in the future this might become a graver problem.
As @superlopuh suggested, we should add a `CHECK` line for each test for anchoring. e.g. 
`// CHECK: pdl.pattern @NonConstantAttrInRewrite` for the first. (I am not sure this is actually printed with the verification failure)